flexmeasures/data/models/planning/tests/test_solver.py

```diff
@@ -2236,3 +2218,272 @@ def test_battery_storage_with_time_series_in_flex_model(
         assert schedule[4:].sum() * 0.25 == pytest.approx(-0.75)
     else:
         assert schedule[4:].sum() * 0.25 == pytest.approx(-0.85)
+
+
+def test_unavoidable_capacity_breach():
+    """Check our ability to limit a capacity breach.
+
+    We want to check two behaviours:
+    1) The breach should be as small as possible (i.e. flattened from the top).
+    2) The breach should not be repeated if that can be avoided (i.e. no taking advantage of created slacks).
+    """
+    start = pd.Timestamp("2020-01-01T00:00:00+01")
+    end = pd.Timestamp("2020-01-02T00:00:00+01")
+    resolution = timedelta(hours=1)
+    soc_at_start = 0.4
+    soc_max = 1
+    soc_min = 0
+    power_capacity = 0.1
+
+    # This target means we must breach the consumption capacity (from 0.1 to 0.2)
+    soc_target = 0.6
+    soc_target_datetime = pd.Timestamp("2020-01-01T01:00:00+01")
+
+    device_constraints = [
+        initialize_df(StorageScheduler.COLUMNS, start, end, resolution)
+    ]
+    ems_constraints = initialize_df(StorageScheduler.COLUMNS, start, end, resolution)
+    empty_commitment = initialize_df(
+        [
+            "quantity",
+            "downwards deviation price",
+            "upwards deviation price",
+            "group",
+        ],
+        start,
+        end,
+        resolution,
+    )
+    commitments = []
+    commitments.append(empty_commitment.copy())
+
+    device_constraints[0]["max"] = soc_max - soc_at_start
+    device_constraints[0]["min"] = soc_min - soc_at_start
+    device_constraints[0]["derivative max"] = 1
+    device_constraints[0]["derivative min"] = -1
+
+    slope = np.arange(0, len(commitments[0])) / (len(commitments[0]) - 1)
+
+    # Introduce an amazing chance to consume without costs
+    slope[10] = -100
+
+    # Day Ahead market commitment
+    commitments[0]["quantity"] = 0
+    commitments[0]["downwards deviation price"] = 90 + slope
+    commitments[0]["upwards deviation price"] = 100 + slope
+    commitments[0]["group"] = list(range(len(commitments[0])))
```

Comment:
Why `len(commitments[0])` - that is the number of keys in that dict, right?

Reply:
`len(commitments[0])` is the number of time slots. Each slot is part of a unique group. I'll add this as an inline comment.

Reply:
Did you add this inline comment?